### ISSUE TITLE
list inside union: java.lang.VerifyError and "Error exists in the bytecode"

### ISSUE THREAD

**Issue by fleminra** [twitter/scrooge#118]
The following compiles fine but has a runtime error when ```f()``` is invoked:
```
union Value
{
  // fails
  1: list<i32> b;
}

service Service
{
  Value f();
}
```

The mere _presence_ of a ```list``` inside the ```union``` causes this problem, even if the server code does not create a ```union``` that uses the field defined as the ```list```.

The error:
```
SEVERE: A server service  threw an exception
java.lang.VerifyError: Uninitialized object exists on backward branch 133
Exception Details:
  Location:
    thrift/Value$.decode(Lorg/apache/thrift/protocol/TProtocol;)Lthrift/Value; @162: goto
  Reason:
    Error exists in the bytecode
  Bytecode:
    0000000: 0157 014d 2bb6 0030 572b b600 334e 2db4
    0000010: 0039 3604 1504 aa00 0000 0012 0000 0001
...
  Stackmap Table:
    append_frame(@40,Null,Object[#53],Integer)
    same_frame(@51)
    append_frame(@76,Integer)
    same_frame(@87)
    full_frame(@116,{Object[#2],Object[#45],Null,Object[#53],Integer,Integer,Object[#75]},{Uninitialized[#87],Uninitialized[#87]})
    full_frame(@133,{Object[#2],Object[#45],Null,Object[#53],Integer,Integer,Object[#75],Object[#89],Integer},{Uninitialized[#87],Uninitialized[#87]})
    full_frame(@165,{Object[#2],Object[#45],Null,Object[#53],Integer,Integer,Object[#75],Object[#89],Integer},{Uninitialized[#87],Uninitialized[#87]})
    full_frame(@171,{Object[#2],Object[#45],Null,Object[#53],Integer,Integer,Object[#75]},{Uninitialized[#87],Uninitialized[#87],Object[#137]})
    full_frame(@175,{Object[#2],Object[#45],Object[#69],Object[#53],Integer},{})
    full_frame(@193,{Object[#2],Object[#45],Object[#69],Object[#53],Integer,Top,Top,Top,Top,Integer,Integer},{})
    same_frame(@217)
    append_frame(@238,Object[#53])
    full_frame(@257,{Object[#2],Object[#45],Object[#69],Object[#53],Integer},{})
    same_frame(@275)

	at thrift.Value$B.write(Value.scala:87)
	at thrift.Service$f$result$class.write(Service.scala:219)
	at thrift.Service$f$result$Immutable.write(Service.scala:194)
	at thrift.Service$FinagledService.reply(Service.scala:366)
	at thrift.Service$FinagledService$$anonfun$1$$anonfun$apply$7.apply(Service.scala:402)
	at thrift.Service$FinagledService$$anonfun$1$$anonfun$apply$7.apply(Service.scala:401)
	at com.twitter.util.Future$$anonfun$flatMap$1.apply(Future.scala:781)
	at com.twitter.util.Future$$anonfun$flatMap$1.apply(Future.scala:780)
	at com.twitter.util.ConstFuture$$anonfun$transform$1.apply(Future.scala:1074)
	at com.twitter.util.ConstFuture$$anonfun$transform$1.apply(Future.scala:1073)
	at com.twitter.util.ConstFuture$$anon$4$$anonfun$run$1.apply$mcV$sp(Future.scala:1063)
	at com.twitter.util.Monitor$.apply(Monitor.scala:128)
	at com.twitter.util.ConstFuture$$anon$4.run(Future.scala:1063)
	at com.twitter.concurrent.LocalScheduler$Activation.run(Scheduler.scala:129)
	at com.twitter.concurrent.LocalScheduler$Activation.submit(Scheduler.scala:107)
	at com.twitter.concurrent.LocalScheduler.submit(Scheduler.scala:150)
	at com.twitter.concurrent.Scheduler$.submit(Scheduler.scala:59)
	at com.twitter.util.Promise.runq(Promise.scala:310)
	at com.twitter.util.Promise.updateIfEmpty(Promise.scala:605)
	at com.twitter.util.Promise.update(Promise.scala:583)
	at com.twitter.util.Promise.setValue(Promise.scala:559)
	at com.twitter.concurrent.AsyncQueue.offer(AsyncQueue.scala:71)
	at com.twitter.finagle.transport.ChannelTransport.handleUpstream(ChannelTransport.scala:41)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboy.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.channel.SimpleChannelHandler.messageReceived(SimpleChannelHandler.java:142)
	at org.jboss.netty.channel.SimpleChannelHandler.handleUpstream(SimpleChannelHandler.java:88)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.jboss.netty.handler.codec.oneone.OneToOneDecoder.handleUpstream(OneToOneDecoder.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at com.twitter.finagle.thrift.ThriftFrameCodec.handleUpstream(ThriftFrameCodec.scala:17)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
	at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
	at org.jboss.netty.channel.socket.nio.AbstractNioSerun(AbstractNioSelector.java:318)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
	at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
	at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:744)

```

Versions:
* com.twitter:finagle-thrift_2.10:6.13.1
* com.twitter:scrooge-runtime_2.10:3.13.0
* org.apache.thrift:libthrift:0.9.1
* org.scala-lang:scala-library:2.10.4
* Oracle Java 1.7.0_51

Here's Server.scala:
```
object Server
{
  def main(args: Array[String])
  {
    import thrift.{Value, Service}
    
    class ServiceImpl extends Service.FutureIface
    {
      import com.twitter.util.Future
    
      def f(): Future[Value] = Future(Value.B(Array(4)))
    }

    val service = new Service.FinagledService(new ServiceImpl,
					      new org.apache.thrift.protocol.TBinaryProtocol.Factory)

    val server = com.twitter.finagle.builder.ServerBuilder()
      .bindTo(new java.net.InetSocketAddress(2000))
      .codec(com.twitter.finagle.thrift.ThriftServerFramedCodec())
      .name("thriftserver")
      .build(service)
  }
}
```

I'm invoking it from CPython + Thrift 0.9.1.

Am I doing anything wrong? I'm somewhat new to Thrift.

Thanks!

**Comment by boclair** [comment]
I'm encountering the same bug, and I also have a union with a list in it.

Using scrooge 3.17, thrift 0.9.2, and sbt 13.6

Any updates?

**Comment by nshkrob** [comment]
The example from the first comment works for me (I've modified https://github.com/twitter/finagle/tree/master/finagle-example to run it). Note that finagle uses libthrift 0.5.0. 
How are you running it and what are you seeing?

**Comment by boclair** [comment]
Thanks for replying.  In summary I fixed my problem by upgrading to JDK version 1.8.0_25.

I believe that scrooge was generating code that was hitting this bug https://issues.scala-lang.org/browse/SI-8645  And thus throwing a
```
java.lang.VerifyError: Uninitialized object exists on backward branch 240
```

I have more info if you are curious, but I believe the issue to be resolved now.  Thanks.

**Comment by nshkrob** [comment]
Great, thanks for digging into this. I'm closing the issue.
